### PR TITLE
patient-data indexing tests

### PR DESCRIPTION
**Pull request by tmarathe**

### CODE REVIEW

**Comment by tmarathe** [comment]
@veronikaslc, @deanway please review!

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import org.phenotips.data.permissions.internal.visibility.PublicVisibility;
+import org.phenotips.vocabulary.SolrCoreContainerHandler;
+import org.slf4j.Logger;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.util.ReflectionUtils;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.query.Query;
+import org.xwiki.query.QueryException;
+import org.xwiki.query.QueryManager;
+import org.xwiki.test.mockito.MockitoComponentMockingRule;
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: [import order](https://github.com/phenotips/build-tools/blob/master/codestyle-resources/src/main/resources/eclipse.importorder) should be fixed

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.junit.Assert;
+
+
+import static org.mockito.Mockito.*;
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: avoid * imports

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java (hunk 1 above)
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: in other files here too

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+        this.server = mock(SolrClient.class);
+
+        this.permissions = this.mocker.getInstance(PermissionsManager.class);
+
+        this.qm = this.mocker.getInstance(QueryManager.class);
+
+        this.patientRepository = this.mocker.getInstance(PatientRepository.class);
+
+        this.patient = mock(PhenoTipsPatient.class);
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: Why are you creating new mock objects and assigning them to these fields? This is already done by the `@Mock` annotations along with `MockitoAnnotations.initMocks(this);` In the cases where you are using `getIntance()` to get mocked fields from the MockitoComponentMockingRule you should remove the `@Mock` annotations

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemoverTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.data.indexing.internal;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.BaseProperty;
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemoverTest.java]: This import is not used

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/PatientEventListenerTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+package org.phenotips.data.indexing.internal;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.phenotips.data.Patient;
+import org.phenotips.data.events.PatientDeletedEvent;
+import org.phenotips.data.events.PatientDeletingEvent;
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/PatientEventListenerTest.java]: This import is not used

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+        this.solrPatientIndexer = (SolrPatientIndexer) this.mocker.getComponentUnderTest();
+        ReflectionUtils.setFieldValue(this.solrPatientIndexer, "server", this.server);
+    }
+
+    @Test
+    public void indexDefaultBehaviourTest() throws IOException, SolrServerException {
+
+        Set<Feature> patientFeatures = new HashSet<>();
+        Feature testFeature = mock(PhenoTipsFeature.class);
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: There is no need to create a mock of an implementation class if only the interface methods are used. `mock(Feature.class)` would be better here

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        this.patientDocReference = new DocumentReference("wiki", "patient", "P0000001");
+
+        doReturn(mock(CoreContainer.class)).when(this.cores).getContainer();
+
+        this.solrPatientIndexer = (SolrPatientIndexer) this.mocker.getComponentUnderTest();
+        ReflectionUtils.setFieldValue(this.solrPatientIndexer, "server", this.server);
+    }
+
+    @Test
+    public void indexDefaultBehaviourTest() throws IOException, SolrServerException {
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: Only the "default" behaviour is tested here, but what about if something goes wrong?
Tests should be added for:
-When patient.getReporter() returns null
-When server.add() throws a SolrServerException
-When server.add() throws an IOException

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        this.solrPatientIndexer.index(this.patient);
+        SolrInputDocument inputDoc = capturedArgument.getLastValue();
+        verify(this.server).add(inputDoc);
+        Assert.assertEquals(inputDoc.getFieldValue("visibility"), "public");
+        Assert.assertEquals(inputDoc.getFieldValue("type"), "id");
+
+    }
+
+    @Test
+    public void deleteDefaultBehaviourTest() throws IOException, SolrServerException {
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: Add tests for exception handling in the delete method, as mention above about index method

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+        doReturn(this.patientDocReference).when(this.patient).getDocument();
+        this.solrPatientIndexer.delete(this.patient);
+        verify(this.server).deleteByQuery("document:"
+                + ClientUtils.escapeQueryChars(this.patientDocReference.toString()));
+        verify(this.server).commit();
+    }
+
+    @Test
+    public void reindexDefaultBehaviour() throws QueryException, IOException, SolrServerException {
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/SolrPatientIndexerTest.java]: Add tests for exception handling of reindex method

File: components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemoverTest.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+        this.patientEmptyObjectsRemover.onEvent(mock(Event.class), xWikiDocument, mock(Object.class));
+        verify(this.xWikiDocument, times(2)).removeXObject((BaseObject) anyObject());
+        verify(this.xWiki, times(2)).saveDocument(this.xWikiDocument, "Removed empty object", true, this.context);
+    }
+
+    @Test
+    public void exitOnNullObjectTest() {
+        this.patientEmptyObjectsRemover.onEvent(mock(Event.class), xWikiDocument, mock(Object.class));
+        verify(this.xWikiDocument, never()).removeXObject((BaseObject) anyObject());
```
Comment (DeanWay) [inline components/patient-data/indexing/src/test/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemoverTest.java]: This test is not very explicit. What are you testing?

**Comment by tmarathe** [comment]
@DeanWay all issues mentioned should be fixed now.

**Pull request merged**